mlflow/gateway/providers/mlflow.py

```diff
@@ -0,0 +1,143 @@
+from fastapi import HTTPException
+from fastapi.encoders import jsonable_encoder
+import json
+
+from .base import BaseProvider
+from .utils import send_request
+from ..config import RouteConfig, MlflowModelServingConfig
+from ..schemas import completions, chat, embeddings
+
+
+class MlflowModelServingProvider(BaseProvider):
+    def __init__(self, config: RouteConfig) -> None:
+        super().__init__(config)
+        if config.model.config is None or not isinstance(
+            config.model.config, MlflowModelServingConfig
+        ):
+            raise TypeError(f"Invalid config type {config.model.config}")
+        self.mlflow_config: MlflowModelServingConfig = config.model.config
+        self.headers = {"Content-Type": "application/json"}
+
+    @staticmethod
+    def process_payload(payload, key):
+        payload = jsonable_encoder(payload, exclude_none=True)
+
+        input_data = payload.pop(key)
+        request_payload = {"inputs": input_data if isinstance(input_data, list) else [input_data]}
+
+        if payload:
+            request_payload["params"] = payload
+
+        return request_payload
+
+    @staticmethod
+    def process_response(response_text):
+        return (
+            response_text if isinstance(response_text, (str, list)) else json.dumps(response_text)
+        )
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        # Example request to MLflow REST API server for completions:
+        # {
+        #     "inputs": ["hi", "hello", "bye"],
+        #     "params": {
+        #         "temperature": 0.5,
+        #         "top_k": 3,
+        #     }
+        # }
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=self.process_payload(payload, "prompt"),
+        )
+
+        # Example response:
+        # {"predictions": ["hello", "hi", "goodbye"]}
+
+        return completions.ResponsePayload(
+            **{
+                "candidates": self.process_response(resp["predictions"]),
```

Comment:
The model's `predictions` field should itself contain a dictionary of the form:

```
{"candidates": ["Completion 1", "completion 2", ...]}
```

By embedding this dictionary under the key `"candidates"`, we're creating:

```
{"candidates": {"candidates": ["completion 1", "completion 2...", ...]}}
```
..
Can we always assume that the model will return a response of the form `{"candidates": ["completion 1", "completion 2", ...]}`, validate the model response to ensure it has this form, and then include the candidates in the final completions response as follos:

```
{"candidates": [{"text": "completion 1", "metadata": {}}, {"text": "completion 2", metadata: {}}]}
```

This way, the response format is consistent with the format used by anthropic, cohere, openai, etc.

Reply:
If only a single string column is defined, using it seems fine too, even if the column name isn't `candidates`. This will help with transformers compatibility until we have standardized signatures for completions in pyfunc. Regardless, we should valid the structure of `resp["predictions"]`.